### ISSUE TITLE
Invalid address, incorrect gem unprocessable entity error handling

### ISSUE THREAD

**Issue by bry** [recurly/recurly-client-ruby#348]
In the recurly-client-ruby version 2.11.2, when using the Recurly `Account#update_attributes!(options)` method and passing an invalid address in the options hash, the Recurly gem incorrectly handles the
```
Recurly::API::UnprocessableEntity Exception: 422 Unprocessable Entity
``` 
error returned from the Recurly API and raises a 
```
NoMethodError Exception: undefined method `invalid!' for #<Hash:0x007fa4e44b46b0>
```

### Input
```ruby
options = {:first_name=>"bryan", :last_name=>"cabalo", :address=>{:address1=>"Flat 6", :address2=>"Green street", :city=>"Elizabeth Court", :state=>"WOR", :country=>"US", :zip=>"DY101JN"}}
```

### StackTrace
```log
Recurly::API::UnprocessableEntity: 422 Unprocessable Entity (PUT https://myproject.recurly.com/v2/accounts/[account id])
1
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/api/net_http_adapter.rb" line 101 in request
2
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/api.rb" line 65 in put
3
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/resource.rb" line 847 in save
4
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/resource.rb" line 873 in save!
5
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/resource.rb" line 920 in update_attributes!
6
File "/app/lib/myproject/recurly/account_service.rb" line 9 in update
7
File "/app/lib/myproject/recurly/shipping_address_service.rb" line 11 in update
8
File "/app/app/workers/shipping_address_worker.rb" line 9 in perform
9
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/processor.rb" line 151 in execute_job
10
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/processor.rb" line 133 in block (2 levels) in process
11
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/middleware/chain.rb" line 127 in block in invoke
12
File "/app/vendor/bundle/ruby/2.2.0/gems/rollbar-2.13.3/lib/rollbar/plugins/sidekiq/plugin.rb" line 11 in call
13
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/middleware/chain.rb" line 129 in block in invoke
14
File "/app/vendor/bundle/ruby/2.2.0/gems/newrelic_rpm-4.4.0.336/lib/new_relic/agent/instrumentation/sidekiq.rb" line 33 in block in call
15
File "/app/vendor/bundle/ruby/2.2.0/gems/newrelic_rpm-4.4.0.336/lib/new_relic/agent/instrumentation/controller_instrumentation.rb" line 362 in perform_action_with_newrelic_trace
16
File "/app/vendor/bundle/ruby/2.2.0/gems/newrelic_rpm-4.4.0.336/lib/new_relic/agent/instrumentation/sidekiq.rb" line 29 in call
17
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/middleware/chain.rb" line 129 in block in invoke
18
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/middleware/server/retry_jobs.rb" line 74 in call
19
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/middleware/chain.rb" line 129 in block in invoke
20
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/middleware/server/logging.rb" line 11 in block in call
21
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/logging.rb" line 31 in with_context
22
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/middleware/server/logging.rb" line 7 in call
23
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/middleware/chain.rb" line 129 in block in invoke
24
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/middleware/chain.rb" line 132 in call
25
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/middleware/chain.rb" line 132 in invoke
26
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/processor.rb" line 128 in block in process
27
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/processor.rb" line 167 in stats
28
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/processor.rb" line 127 in process
29
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/processor.rb" line 79 in process_one
30
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/processor.rb" line 67 in run
31
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/util.rb" line 16 in watchdog
32
File "/app/vendor/bundle/ruby/2.2.0/gems/sidekiq-4.1.1/lib/sidekiq/util.rb" line 24 in block in safe_thread
NoMethodError: undefined method `invalid!' for #<Hash:0x007f5149086b40>
1
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/resource.rb" line 1050 in invalid!
2
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/resource.rb" line 1030 in block in apply_errors
3
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/xml/nokogiri.rb" line 16 in block in each_element
4
File "/app/vendor/bundle/ruby/2.2.0/gems/nokogiri-1.6.8.1/lib/nokogiri/xml/node_set.rb" line 187 in block in each
5
File "/app/vendor/bundle/ruby/2.2.0/gems/nokogiri-1.6.8.1/lib/nokogiri/xml/node_set.rb" line 186 in upto
6
File "/app/vendor/bundle/ruby/2.2.0/gems/nokogiri-1.6.8.1/lib/nokogiri/xml/node_set.rb" line 186 in each
7
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/xml/nokogiri.rb" line 16 in each_element
8
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/xml.rb" line 67 in each_element
9
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/resource.rb" line 1028 in apply_errors
10
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/resource.rb" line 855 in rescue in save
11
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/resource.rb" line 845 in save
12
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/resource.rb" line 873 in save!
13
File "/app/vendor/bundle/ruby/2.2.0/gems/recurly-2.11.2/lib/recurly/resource.rb" line 920 in update_attributes!
14
File "/app/lib/myproject/recurly/account_service.rb" line 9 in update
```

**Comment by bry** [comment]
@bhelx The explicit `Recurly::Address` casting works and the bug in the gem is gone, Recurly now reports with its `Recurly::Resource::Invalid` object. 👍 

Thanks for looking into this!

**Comment by bhelx** [comment]
@bry glad to hear. I'll still keep this open in case we get time to fix the general case.

**Comment by aaron-suarez** [comment]
Is there any chance #455 fixed the general case for this? @bhelx Were you able to reproduce this? Perhaps we can verify and close this out

**Comment by bhelx** [comment]
I think it's safe to close with #455 . Please re-open @bry if you feel like it wasn't addressed.

**Comment by aaron-suarez** [comment]
@bry The fix from #455 was released just now to rubygems, I appreciate you re-opening this if that fix does not solve this problem. Thanks!
